File: solutions/cpp/atbash-cipher/1/atbash_cipher.cpp
#include "atbash_cipher.h"
#include <string>
#include <map>
#include <vector>

namespace atbash_cipher {
    const std::string key_alphabet    = "abcdefghijklmnopqrstuvwxyz";
    const std::string cipher_alphabet = "zyxwvutsrqponmlkjihgfedcba";

    class grouped_string {
        public: 
            grouped_string(unsigned int size, char sep = ' ') : _buffer{}, _sep{sep}, _size{size} {}
            void push_back(char c) {
                _buffer.push_back(c);
                _chars++;                
                if(_chars > 0 && _chars % _size == 0) {
                    _buffer.push_back(_sep);
                }
  
            }
            std::string to_string() {
                if(_buffer.back() == _sep) _buffer.pop_back();
                return std::string(_buffer.begin(), _buffer.end());
            }
            ~grouped_string() {
                _buffer.clear();
            }
        private:
            std::vector<char> _buffer;
            char _sep;
            unsigned int _chars = 0;
            unsigned int _size;
    };
    
    std::string encode(const std::string& plaintext) {
        grouped_string raw_result{5};
        for(auto c : plaintext) {
            if(std::isdigit(c)) {
                raw_result.push_back(c);
            }
            else {
                auto i = key_alphabet.find(std::tolower(c));
                if(i != std::string::npos) {
                    raw_result.push_back(cipher_alphabet.at(i));
                }
            }
        }
        return raw_result.to_string();
    }

    std::string decode(const std::string& ciphertext) {
        std::string plaintext{};
        for(auto c : ciphertext) {
            if(std::isdigit(c)) {
                plaintext.push_back(c);
            }
            else {
                auto i = cipher_alphabet.find(c);
                if(i != std::string::npos)  {
                    plaintext.push_back(key_alphabet.at(i));
                }
            }
        }
        return plaintext;
    }
}  // namespace atbash_cipher

File: solutions/dart/atbash-cipher/1/lib/atbash_cipher.dart
class AtbashCipher {
  List<String> plaintext;
  List<String> ciphertext;
  
  AtbashCipher({String keytext = "abcdefghijklmnopqrstuvwxyz"})
    : 
    this.plaintext = keytext.split(''),
    this.ciphertext = keytext.split('').reversed.toList();
  
  String encode(String message) {
    final _message = message.toLowerCase().split('');
    var _result = [];
    var _c = 0;
    for(var n=0; n < _message.length; n++) {
      final l = _message[n];
      if(l == " " || l == "." || l == ",") continue;
      final i = this.plaintext.indexOf(l);
      if(i == -1) {
        _result.add(l);
      }
      else {
        _result.add(ciphertext[i]);
      }
      _c += 1;
      if(_c % 5 == 0) {
        _result.add(' ');
      }
    }
    return _result.join('').trim();
  }

  String decode(String message) {
    var _result = [];
    for(var n=0; n < message.length; n++) {
      final l = message[n];
      if(l == " ") continue;
      final i = ciphertext.indexOf(l);
      if(i == -1) {
        _result.add(l);
      }
      else {
        _result.add(plaintext[i]);
      }
    } 
    return _result.join('');
  }
}
